core/chains/cosmos/chain.go

```diff
@@ -0,0 +1,168 @@
+// however there's nothing we can do but wait until the block is processed.
+// So we set a fairly high timeout here.
+const DefaultRequestTimeout = 30 * time.Second
+
+//go:generate mockery --quiet --name TxManager --srcpkg github.com/smartcontractkit/chainlink-cosmos/pkg/cosmos/adapters --output ./mocks/ --case=underscore
+//go:generate mockery --quiet --name Reader --srcpkg github.com/smartcontractkit/chainlink-cosmos/pkg/cosmos/client --output ./mocks/ --case=underscore
+//go:generate mockery --quiet --name Chain --srcpkg github.com/smartcontractkit/chainlink-cosmos/pkg/cosmos/adapters --output ./mocks/ --case=underscore
+var _ adapters.Chain = (*chain)(nil)
+
+type chain struct {
+	utils.StartStopOnce
+	id           string
+	cfg          coscfg.Config
+	cfgImmutable bool // toml config is immutable
+	txm          *cosmostxm.Txm
+	orm          types.ORM
+	lggr         logger.Logger
+}
+
+func newChain(id string, cfg coscfg.Config, db *sqlx.DB, ks keystore.Cosmos, logCfg pg.QConfig, eb pg.EventBroadcaster, orm types.ORM, lggr logger.Logger) (*chain, error) {
+	lggr = lggr.With("cosmosChainID", id)
+	var ch = chain{
+		id:   id,
+		cfg:  cfg,
+		orm:  orm,
+		lggr: lggr.Named("Chain"),
+	}
+	tc := func() (cosmosclient.ReaderWriter, error) {
+		return ch.getClient("")
+	}
+	gpe := cosmosclient.NewMustGasPriceEstimator([]cosmosclient.GasPricesEstimator{
+		cosmosclient.NewClosureGasPriceEstimator(func() (map[string]sdk.DecCoin, error) {
+			return map[string]sdk.DecCoin{
+				"uatom": sdk.NewDecCoinFromDec("uatom", cfg.FallbackGasPriceUAtom()),
+			}, nil
+		}),
+	}, lggr)
+	ch.txm = cosmostxm.NewTxm(db, tc, *gpe, ch.id, cfg, ks, lggr, logCfg, eb)
+
+	return &ch, nil
+}
+
+func (c *chain) Name() string {
+	return c.lggr.Name()
+}
+
+func (c *chain) ID() string {
+	return c.id
+}
+
+func (c *chain) Config() coscfg.Config {
+	return c.cfg
+}
+
+func (c *chain) UpdateConfig(cfg *db.ChainCfg) {
+	if c.cfgImmutable {
+		c.lggr.Criticalw("TOML configuration cannot be updated", "err", v2.ErrUnsupported)
+		return
+	}
+	c.cfg.Update(*cfg)
+}
+
+func (c *chain) TxManager() adapters.TxManager {
+	return c.txm
+}
+
+func (c *chain) Reader(name string) (cosmosclient.Reader, error) {
+	return c.getClient(name)
+}
+
+// getClient returns a client, optionally requiring a specific node by name.
+func (c *chain) getClient(name string) (cosmosclient.ReaderWriter, error) {
+	//TODO cache clients?
+	var node db.Node
+	if name == "" { // Any node
+		nodes, cnt, err := c.orm.NodesForChain(c.id, 0, math.MaxInt)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to get nodes")
+		}
+		if cnt == 0 {
+			return nil, errors.New("no nodes available")
+		}
+		// #nosec
+		node = nodes[rand.Intn(len(nodes))]
+	} else { // Named node
+		var err error
+		node, err = c.orm.NodeNamed(name)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get node named %s", name)
+		}
+		if node.CosmosChainID != c.id {
```

Comment:
I think we should move this check to the orm method and require chain ID as a parameter to `NodeNamed`

Reply:
I think that would be special casing this method a little too much, since it is not always used with that limitation. It's also generic and shared across chain families.